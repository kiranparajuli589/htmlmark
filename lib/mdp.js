import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
