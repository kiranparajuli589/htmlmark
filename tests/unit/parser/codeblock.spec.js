import {MDP} from "../../mdp.js"


const mdp = new MDP()


describe("CodeBlock Parsing", () => {
	it("should parse the codeblock", () => {
		const lines = [
			"```js",
			"const a = 1",
			"```"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse the codeblock without the language", () => {
		const lines = [
			"```",
			"const a = 1",
			"```"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse the codeblock with multiple lines", () => {
		const lines = [
			"```js",
			"const a = 1",
			"const b = 2",
			"```"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse the codeblock with multiple lines and a newline", () => {
		const lines = [
			"```js",
			"const a = 1",
			"const b = 2",
			"",
			"const c = 3",
			"```"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse multiple consequetive codeblocks", () => {
		const lines = [
			"```js",
			"const a = 1",
			"```",
			"```js",
			"const b = 2",
			"```"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should cope with multiple items", () => {
		const lines = [
			"```js",
			"const a = 1",
			"```",
			"some people are funny"
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
	it("should parse indent codeblock", () => {
		const lines = [
			"This is a normal paragraph:",
			"",
			"    This is a code block.",
			"",
			"Here is an example of AppleScript:",
			"",
			"    tell application \"Foo\"",
			"    beep",
			"    end tell",
			"",
			"A code block continues until it reaches a line that is not indented",
			"(or the end of the article)."
		]
		const html = mdp.h(lines)
		expect(html).toMatchSnapshot()
	})
})
