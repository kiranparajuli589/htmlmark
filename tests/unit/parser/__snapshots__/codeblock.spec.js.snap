// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBlock Parsing should cope with multiple items 1`] = `
"<pre><code class='language-js'>const a = 1</code></pre>
<p>some people are funny</p>
"
`;

exports[`CodeBlock Parsing should parse indent codeblock 1`] = `
"<p>This is a normal paragraph:</p>
<pre><code>This is a code block.
</code></pre>
<p>Here is an example of AppleScript:</p>
<pre><code>tell application &quot;Foo&quot;
beep
end tell
</code></pre>
<p>A code block continues until it reaches a line that is not indented (or the end of the article).</p>
"
`;

exports[`CodeBlock Parsing should parse multiple consequetive codeblocks 1`] = `
"<pre><code class='language-js'>const a = 1</code></pre>
<pre><code class='language-js'>const b = 2</code></pre>
"
`;

exports[`CodeBlock Parsing should parse the codeblock 1`] = `
"<pre><code class='language-js'>const a = 1</code></pre>
"
`;

exports[`CodeBlock Parsing should parse the codeblock with multiple lines 1`] = `
"<pre><code class='language-js'>const a = 1
const b = 2</code></pre>
"
`;

exports[`CodeBlock Parsing should parse the codeblock with multiple lines and a newline 1`] = `
"<pre><code class='language-js'>const a = 1
const b = 2

const c = 3</code></pre>
"
`;

exports[`CodeBlock Parsing should parse the codeblock without the language 1`] = `
"<pre><code>const a = 1</code></pre>
"
`;
